diff --git a/linux/trampoline.h b/linux/trampoline.h
index 4549459204..ebe9ed2939 100644
--- a/linux/trampoline.h
+++ b/linux/trampoline.h
@@ -2,6 +2,10 @@
 
 #include "linux.h"
 
+#ifndef __STDC_NO_THREADS__
 #include <threads.h>
+#else
+#include "threads.h"
+#endif
 
 extern thread_local struct LFIContext *new_ctx;
diff --git a/core/core.h b/core/core.h
index 5056e9acda..04389eb037 100644
--- a/core/core.h
+++ b/core/core.h
@@ -9,9 +9,13 @@
 #include "mmap.h"
 
 #include <signal.h>
+#ifndef __STDC_NO_THREADS__
 #include <threads.h>
+#else
+#include "threads.h"
+#endif
 
-#define EXPORT __attribute__((visibility("default")))
+#define EXPORT
diff --git a/linux/host.c b/linux/host.c
index 7d5d971ea0..9b7c650459 100644
--- a/linux/host.c
+++ b/linux/host.c
@@ -8,7 +8,6 @@
 #include <assert.h>
 #include <errno.h>
 #include <fcntl.h>
-#include <sys/random.h>
 #include <sys/stat.h>
 #include <unistd.h>
 
@@ -58,12 +57,33 @@ randomflags(unsigned int flags)
         f |= GRND_NONBLOCK;
     return f;
 }
+#elif defined(HAVE_SYS_GETRANDOM)
+/* Flags for use with getrandom.  */
+#define GRND_NONBLOCK 0x01
+#define GRND_RANDOM 0x02
+
+static unsigned int
+randomflags(unsigned int flags)
+{
+    unsigned int f = 0;
+    if ((flags & LINUX_GRND_RANDOM) != 0)
+        f |= GRND_RANDOM;
+    if ((flags & LINUX_GRND_NONBLOCK) != 0)
+        f |= GRND_NONBLOCK;
+    return f;
+}
+
+#define SYS_getrandom 318
+
+static ssize_t getrandom(void* buf, size_t buflen, unsigned int flags) {
+    return syscall(SYS_getrandom, buf, buflen, flags);
+}
 #endif
 
 ssize_t
 host_getrandom(void *buf, size_t size, unsigned int flags)
 {
-#if defined(HAVE_GETRANDOM)
+#if defined(HAVE_GETRANDOM) || defined(HAVE_SYS_GETRANDOM)
     ssize_t r = getrandom(buf, size, randomflags(flags));
     if (r < 0)
         return host_err(errno);
diff --git a/linux/linux.h b/linux/linux.h
index 4ac2ecaa96..493ca9fbc2 100644
--- a/linux/linux.h
+++ b/linux/linux.h
@@ -4,7 +4,7 @@
 #include "lfi_linux.h"
 #include "log.h"
 
-#define EXPORT __attribute__((visibility("default")))
+#define EXPORT
 
 struct LFILinuxEngine {
     struct LFIEngine *engine;
