diff --git a/liblfi/host/linux/random.c b/liblfi/host/linux/random.c
index fc3ca368c7..086ee5eaca 100644
--- a/liblfi/host/linux/random.c
+++ b/liblfi/host/linux/random.c
@@ -1,9 +1,19 @@
 #include <errno.h>
-#include <sys/random.h>
+#include <unistd.h>
 
 #include "host/error.h"
 #include "host.h"
 
+#define SYS_getrandom 318
+
+/* Flags for use with getrandom.  */
+#define GRND_NONBLOCK 0x01
+#define GRND_RANDOM 0x02
+
+static ssize_t getrandom(void* buf, size_t buflen, unsigned int flags) {
+    return syscall(SYS_getrandom, buf, buflen, flags);
+}
+
 static unsigned int
 linuxflags(unsigned int flags)
 {
diff --git a/liblfi/types.h b/liblfi/types.h
index b5a51e65c7..fa1e0d7b78 100644
--- a/liblfi/types.h
+++ b/liblfi/types.h
@@ -2,7 +2,7 @@
 
 #include <stdint.h>
 
-#define EXPORT __attribute__((visibility("default")))
+#define EXPORT
 
 typedef uint64_t  u64;
 typedef uint32_t  u32;
