diff --git a/liblfi/elfload.c b/liblfi/elfload.c
index 721e504f97..85990756c0 100644
--- a/liblfi/elfload.c
+++ b/liblfi/elfload.c
@@ -45,7 +45,7 @@ sanitize(void* p, size_t sz, int prot)
 {
     if ((prot & LFI_PROT_EXEC) == 0)
         return;
-#if (defined(__x86_64__) || defined(_M_X64)) && TUX_SANITIZE_CODE
+#if (defined(__x86_64__) || defined(_M_X64)) && JS_SANDBOX_VERIFY
     const uint8_t SAFE_BYTE = 0xcc;
     memset(p, SAFE_BYTE, sz);
 #endif
diff --git a/liblfi/pal/platform.c b/liblfi/pal/platform.c
index f98f05076d..5520320af5 100644
--- a/liblfi/pal/platform.c
+++ b/liblfi/pal/platform.c
@@ -15,9 +15,13 @@ static void showerr(char* msg, size_t sz) {
 EXPORT struct LFIPlatform*
 lfi_new_plat(struct LFIPlatOptions opts)
 {
+#ifdef JS_SANDBOX_VERIFY
     struct LFIVerifier* verifier = malloc(sizeof(LFIVerifier));
     if (!verifier)
         return NULL;
+#else
+    struct LFIVerifier* verifier = NULL;
+#endif
 
     struct LFIPlatform* plat = malloc(sizeof(struct LFIPlatform));
     if (!plat)
@@ -32,7 +36,7 @@ lfi_new_plat(struct LFIPlatOptions opts)
         goto err2;
     if (!boxmap_reserve(bm, gb(512)))
         goto err3;
-
+#ifdef JS_SANDBOX_VERIFY
     verifier->opts = (struct LFIVOptions) {
         .box = LFI_BOX_STORES,
         .guardsize = gb(2),
@@ -40,6 +44,7 @@ lfi_new_plat(struct LFIPlatOptions opts)
     };
 
     verifier->verify = lfiv_verify_x64_large;
+#endif
 
     *plat = (struct LFIPlatform) {
         .bm = bm,
@@ -53,7 +58,9 @@ err3:
 err2:
     free(plat);
 err1:
+#ifdef JS_SANDBOX_VERIFY
     free(verifier);
+#endif
     return NULL;
 }
diff --git a/liblfi/syscalls/sys_jitcode.c b/liblfi/syscalls/sys_jitcode.c
index 1dc5e86414..90efaabcc2 100644
--- a/liblfi/syscalls/sys_jitcode.c
+++ b/liblfi/syscalls/sys_jitcode.c
@@ -23,7 +23,6 @@ uintptr_t sys_jitcode_mmap(struct TuxProc* p, lfiptr_t addrp, size_t exec_length
 }

 int sys_jitcode_create(struct TuxProc* p, lfiptr_t addrp, lfiptr_t bufp, size_t length) {
+#ifdef JS_SANDBOX_VERIFY
   // Make sure addresses are bundle aligned
   int bundle_size = 32;

@@ -36,7 +35,6 @@ int sys_jitcode_create(struct TuxProc* p, lfiptr_t addrp, lfiptr_t bufp, size_t
       VERBOSE(p->tux, "sys_jitcode_create: length not bundle aligned!");
       return -1;
   }
+#endif

   uint8_t* src = procbuf(p, bufp, length);
   uint8_t* buf = (uint8_t*)malloc(length);
@@ -51,7 +49,6 @@ int sys_jitcode_create2(struct TuxProc *p, lfiptr_t addrp, lfiptr_t headerp,
                         size_t total_length) {
   // TODO: We probably need better sanity checks here
   assert(total_length >= header_length);
+#ifdef JS_SANDBOX_VERIFY
   // Make sure addresses are bundle aligned
   int bundle_size = 32;

@@ -69,7 +66,6 @@ int sys_jitcode_create2(struct TuxProc *p, lfiptr_t addrp, lfiptr_t headerp,
       VERBOSE(p->tux, "sys_jitcode_create: header_length not bundle aligned!");
       return -1;
   }
+#endif

   uint8_t* header = procbuf(p, headerp, header_length);
