diff --git a/liblfi/lfi.h b/liblfi/lfi.h
index 2812601e7a..f01efbc4fa 100644
--- a/liblfi/lfi.h
+++ b/liblfi/lfi.h
@@ -5,7 +5,7 @@
 #include <stdint.h>
 #include <sys/types.h>
 
-#include "lfiv.h"
+//#include "lfiv.h"
 
 enum {
     LFI_ERR_OK            = 0,
@@ -27,6 +27,10 @@ enum {
 
 typedef uint64_t (*SysHandler)(void* ctxp, uint64_t sysno, uint64_t, uint64_t, uint64_t, uint64_t, uint64_t, uint64_t);
 
+typedef struct {
+    void* verify;
+} LFIVerifier;
+
 typedef struct {
     // Do not run verification.
     bool noverify;
diff --git a/liblfi/proc.c b/liblfi/proc.c
index 3675b5c113..9331d1c426 100644
--- a/liblfi/proc.c
+++ b/liblfi/proc.c
@@ -8,10 +8,11 @@
 #include <errno.h>
 #include <fcntl.h>
 #include <sys/mman.h>
+#include <syscall.h>
 
 #include "align.h"
 #include "lfi.h"
-#include "lfiv.h"
+//#include "lfiv.h"
 #include "engine.h"
 #include "proc.h"
 #include "err.h"
@@ -21,7 +22,7 @@
 #if defined(__aarch64__) || defined(_M_ARM64)
 #include "arch/arm64/arm64.h"
 #elif defined(__x86_64__) || defined(_M_X64)
-#include "arch/amd64/amd64.h"
+#include "./arch/amd64/amd64.h"
 #endif
 
 extern uint64_t lfi_proc_entry(LFIProc* proc, void** kstackp) asm ("lfi_proc_entry");
@@ -73,6 +74,10 @@ proc_validate(LFIProc* proc)
         *lfi_regs_mask(&proc->regs) = mask(proc->lfi->opts.p2size);
 }
 
+static inline int memfd_create(const char* name, unsigned int flags) {
+    return syscall(__NR_memfd_create, name, flags);
+}
+
 bool
 lfi_uproc_init(LFIProc* proc, uintptr_t codebase, size_t codesize, size_t datasize)
 {
@@ -253,10 +258,10 @@ mprotectverify(void* base, size_t size, int prot, LFIVerifier* verifier)
 {
     if ((prot & PROT_EXEC) == 0 || !verifier)
         return mprotect(base, size, prot);
-    if (!lfiv_verify(verifier, base, size, (uintptr_t) base)) {
-        lfi_errno = LFI_ERR_VERIFY;
-        return -1;
-    }
+    //if (!lfiv_verify(verifier, base, size, (uintptr_t) base)) {
+    //    lfi_errno = LFI_ERR_VERIFY;
+    //    return -1;
+    //}
     return mprotect(base, size, prot);
 }
 
@@ -339,13 +344,13 @@ ureadelfseg(LFIProc* proc, uintptr_t start, uintptr_t offset, uintptr_t end,
             return false;
         }
 
-        LFIVerifier* verifier = proc->lfi->opts.verifier;
-        if (verifier) {
-            if (!lfiv_verify(verifier, &buf.data[p_offset], filesz, start + offset)) {
-                lfi_errno = LFI_ERR_VERIFY;
-                return false;
-            }
-        }
+        //LFIVerifier* verifier = proc->lfi->opts.verifier;
+        //if (verifier) {
+        //    if (!lfiv_verify(verifier, &buf.data[p_offset], filesz, start + offset)) {
+        //        lfi_errno = LFI_ERR_VERIFY;
+        //        return false;
+        //    }
+        //}
         target = proc->ucodealias + (start - proc->ucodebase) + offset;
     } else if (prot == (PROT_READ | PROT_WRITE) || prot == PROT_READ) {
         // Note: for micro processes, read-only segments are mapped read-write.
@@ -662,7 +667,8 @@ procmap(LFIProc* proc, uintptr_t start, size_t size, int prot, int flags, int fd
     assert(start % proc->lfi->opts.pagesize == 0);
     assert(size % proc->lfi->opts.pagesize == 0);
 
-    void* mem = mmapverify((void*) start, size, prot, flags | MAP_FIXED, fd, offset, proc->lfi->opts.verifier);
+    //void* mem = mmapverify((void*) start, size, prot, flags | MAP_FIXED, fd, offset, proc->lfi->opts.verifier);
+    void* mem = mmapverify((void*) start, size, prot, flags | MAP_FIXED, fd, offset, NULL);
     if (mem == (void*) -1)
         return -errno;
     return 0;
@@ -715,7 +721,8 @@ lfi_proc_mprotect(LFIProc* proc, uintptr_t addr, size_t size, int prot)
         return -1;
 
     assert(addr >= proc->base && addr < proc->base + proc->size);
-    return mprotectverify((void*) addr, size, prot, proc->lfi->opts.verifier);
+    //return mprotectverify((void*) addr, size, prot, proc->lfi->opts.verifier);
+    return mprotectverify((void*) addr, size, prot, NULL);
 }
 
 int
diff --git a/liblfix/arch/amd64/syscall.c b/liblfix/arch/amd64/syscall.c
index 379e185563..d56f162369 100644
--- a/liblfix/arch/amd64/syscall.c
+++ b/liblfix/arch/amd64/syscall.c
@@ -1,8 +1,8 @@
 // For mremap
 #define _GNU_SOURCE
 
-#include "syscall.h"
-#include "sys.h"
+#include "../../syscall.h"
+#include "../../sys.h"
 
 enum {
     LSYS_getcwd            = 79,
diff --git a/liblfix/file.c b/liblfix/file.c
index d38f989b63..325dd41824 100644
--- a/liblfix/file.c
+++ b/liblfix/file.c
@@ -4,11 +4,21 @@
 #include <unistd.h>
 #include <stdlib.h>
 #include <dirent.h>
+#include <syscall.h>
+#include <sys/stat.h>
 
 #include "file.h"
 #include "sys.h"
 #include "io.h"
 
+static inline int memfd_create(const char* name, unsigned int flags) {
+    return syscall(__NR_memfd_create, name, flags);
+}
+
+static inline ssize_t getdents64(int fd, void* dirp, size_t count) {
+    return syscall(__NR_getdents64, fd, dirp, count);
+}
+
 FDFile*
 lfix_filenew(LFIXEngine* lfix, int dirfd, const char* name, int flags, int mode)
 {
diff --git a/liblfix/syscall.h b/liblfix/syscall.h
index 006b160611..cabff33f63 100644
--- a/liblfix/syscall.h
+++ b/liblfix/syscall.h
@@ -4,12 +4,12 @@
 #include <stddef.h>
 #include <stdio.h>
 #include <stdalign.h>
+#include <string.h>
 #include <fcntl.h>
 #include <unistd.h>
 #include <time.h>
 #include <sched.h>
 #include <sys/mman.h>
-#include <sys/random.h>
 #include <sys/sysinfo.h>
 #include <sys/syscall.h>
 #include <linux/futex.h>
@@ -22,6 +22,16 @@
 
 #include "syswrap.h"
 
+#define SYS_getrandom 318
+
+static ssize_t getrandom(void* buf, size_t buflen, unsigned int flags) {
+    return syscall(SYS_getrandom, buf, buflen, flags);
+}
+
+static int renameat2(int olddirfd, const char* oldpath, int newdirfd, const char* newpath, unsigned int flags) {
+    return syscall(SYS_renameat2, olddirfd, oldpath, newdirfd, newpath, flags);
+}
+
 static uintptr_t
 truncp(uintptr_t addr, size_t align)
 {
@@ -63,7 +73,7 @@ procbuf(LFIXProc* p, uintptr_t buf, size_t size)
 }
 
 enum {
-    PATH_MAX = 4096,
+    LFIX_PATH_MAX = 4096,
 };
 
 static const char*
@@ -71,7 +81,7 @@ procpath(LFIXProc* p, uintptr_t path)
 {
     path = procaddr(p, path);
     const char* str = (const char*) path;
-    size_t len = strnlen(str, PATH_MAX);
+    size_t len = strnlen(str, LFIX_PATH_MAX);
     if (path + len >= p->base + p->size)
         return NULL;
     if (str[len] != 0)
